File: api/_lib/gemini.py
import json
import ssl
from urllib import error as urllib_error, request as urllib_request

from .runtime import (
    ApiError,
    build_ssl_context,
    get_gemini_api_key,
    get_gemini_fallback_model,
    get_gemini_primary_model,
)


def build_parse_prompt(user_input, current_trackers):
    current_json = json.dumps(current_trackers, ensure_ascii=True, indent=2)
    return f"""
You convert habit-tracker instructions into the final tracker list for a productivity app.

Return only a JSON array. Do not include markdown, code fences, or explanations.

The array must represent the full final tracker list after applying the user's instruction to the current trackers.
If the user wants to add trackers, include them.
If the user wants to rename, modify, or remove trackers, update the array accordingly.
Preserve existing tracker ids when modifying an existing tracker.

Each tracker object should follow these rules:
- id: preserve the existing tracker id when updating an existing tracker, otherwise omit it
- name: short string
- type: one of "binary", "numeric", "session"
- category: one of "study", "diet", "exercise", "habits", "none"
- frequency: "daily" or "weekly"
- logging_mode: "simple" for binary, "quantity" for numeric, "time" for session
- unit: short string, empty string, or null
- goal: number or null
- increments: array of 0 to 3 positive numbers
- primary_action: short verb phrase
- optional_actions: array of up to 3 short strings
- fields: array of field objects with shape {{ "name": string, "type": "number" | "boolean" | "time", "unit": string | null }}

Tracker guidance:
- binary trackers usually use fields like [{{"name":"done","type":"boolean","unit":null}}]
- numeric trackers usually use one number field
- session trackers usually use [{{"name":"duration","type":"time","unit":"minutes"}}]
- weekly trackers should still keep the same general shape

Current trackers:
{current_json}

User instruction:
{user_input}
""".strip()


def build_response_schema():
    return {
        "type": "array",
        "items": {
            "type": "object",
            "properties": {
                "id": {"type": "string"},
                "name": {"type": "string"},
                "type": {"type": "string"},
                "mode": {"type": "string"},
                "category": {"type": "string"},
                "logging_mode": {"type": "string"},
                "unit": {"type": ["string", "null"]},
                "goal": {"type": ["number", "null"]},
                "frequency": {"type": "string"},
                "increments": {
                    "type": "array",
                    "items": {"type": "number"},
                },
                "primary_action": {"type": "string"},
                "optional_actions": {
                    "type": "array",
                    "items": {"type": "string"},
                },
                "fields": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "name": {"type": "string"},
                            "type": {"type": "string"},
                            "unit": {"type": ["string", "null"]},
                        },
                        "required": ["name", "type", "unit"],
                    },
                },
            },
            "required": [
                "name",
                "type",
                "category",
                "logging_mode",
                "frequency",
                "increments",
                "primary_action",
                "optional_actions",
                "fields",
            ],
        },
    }


def strip_code_fences(text):
    cleaned = (text or "").strip()
    if cleaned.startswith("```"):
        cleaned = cleaned.split("\n", 1)[1] if "\n" in cleaned else ""
    if cleaned.endswith("```"):
        cleaned = cleaned[:-3]
    return cleaned.strip()


def extract_balanced_json_snippet(text, open_char, close_char):
    start = text.find(open_char)
    if start == -1:
        return ""

    depth = 0
    in_string = False
    escaped = False

    for index in range(start, len(text)):
        char = text[index]

        if in_string:
            if escaped:
                escaped = False
            elif char == "\\":
                escaped = True
            elif char == '"':
                in_string = False
            continue

        if char == '"':
            in_string = True
            continue

        if char == open_char:
            depth += 1
        elif char == close_char:
            depth -= 1
            if depth == 0:
                return text[start : index + 1]

    return ""


def looks_like_tracker_object(value):
    return isinstance(value, dict) and any(
        isinstance(value.get(key), expected_type)
        for key, expected_type in (
            ("name", str),
            ("type", str),
            ("fields", list),
            ("frequency", str),
        )
    )


def coerce_tracker_array(value, current_tracker_count=0):
    if isinstance(value, list):
        return [tracker for tracker in value if isinstance(tracker, dict)]

    if not isinstance(value, dict):
        return None

    trackers = value.get("trackers")
    if isinstance(trackers, list):
        return [tracker for tracker in trackers if isinstance(tracker, dict)]

    for key in ("result", "data", "items", "output"):
        nested = value.get(key)
        if isinstance(nested, list):
            return [tracker for tracker in nested if isinstance(tracker, dict)]

    if current_tracker_count == 0 and looks_like_tracker_object(value):
        return [value]

    return None


def summarize_gemini_text(text):
    return " ".join(str(text or "").split())[:220]


def parse_trackers_text(text, current_tracker_count=0):
    cleaned = strip_code_fences(text)
    candidates = [cleaned]
    array_snippet = extract_balanced_json_snippet(cleaned, "[", "]")
    object_snippet = extract_balanced_json_snippet(cleaned, "{", "}")

    if array_snippet and array_snippet != cleaned:
        candidates.append(array_snippet)
    if object_snippet and object_snippet not in (cleaned, array_snippet):
        candidates.append(object_snippet)

    for candidate in candidates:
        if not candidate:
            continue
        try:
            parsed = json.loads(candidate)
            if isinstance(parsed, str):
                parsed = json.loads(parsed)
        except json.JSONDecodeError:
            continue

        trackers = coerce_tracker_array(parsed, current_tracker_count=current_tracker_count)
        if trackers is not None:
            return trackers

    raise ApiError(
        f"Gemini model did not return valid JSON array. Received: {summarize_gemini_text(cleaned)}",
        status=502,
        code="AI_INVALID_JSON",
        retryable=True,
    )


def extract_candidate_text(response_payload):
    candidates = response_payload.get("candidates") or []
    if not candidates:
        prompt_feedback = response_payload.get("promptFeedback") or {}
        block_reason = prompt_feedback.get("blockReason")
        if block_reason:
            raise ApiError(
                f"Gemini blocked the prompt: {block_reason}",
                status=422,
                code="AI_BLOCKED",
                retryable=False,
            )
        raise ApiError("Gemini returned no candidates.", status=502, code="AI_EMPTY", retryable=True)

    parts = candidates[0].get("content", {}).get("parts", [])
    text = "".join(part.get("text", "") for part in parts if isinstance(part, dict)).strip()
    if not text:
        raise ApiError("Gemini returned an empty response.", status=502, code="AI_EMPTY", retryable=True)
    return text


def build_payload(user_input, current_trackers):
    payload = {
        "contents": [
            {
                "parts": [
                    {
                        "text": build_parse_prompt(
                            user_input=user_input.strip(),
                            current_trackers=current_trackers if isinstance(current_trackers, list) else [],
                        )
                    }
                ]
            }
        ],
        "generationConfig": {
            "temperature": 0.2,
            "responseMimeType": "application/json",
            "responseJsonSchema": build_response_schema(),
        },
    }
    return payload


def request_gemini(payload, model_name, current_tracker_count=0):
    req = urllib_request.Request(
        f"https://generativelanguage.googleapis.com/v1beta/models/{model_name}:generateContent",
        data=json.dumps(payload).encode("utf-8"),
        headers={
            "Content-Type": "application/json",
            "x-goog-api-key": get_gemini_api_key(),
        },
        method="POST",
    )

    try:
        with urllib_request.urlopen(req, timeout=45, context=build_ssl_context()) as response:
            response_payload = json.loads(response.read().decode("utf-8"))
    except urllib_error.HTTPError as exc:
        details = exc.read().decode("utf-8", errors="ignore")
        raise ApiError(
            f"Gemini API error from {model_name} ({exc.code}): {details or exc.reason}",
            status=502,
            code="AI_UPSTREAM_ERROR",
            retryable=exc.code >= 500,
        ) from exc
    except ssl.SSLCertVerificationError as exc:
        raise ApiError(
            f"TLS certificate verification failed while connecting to Gemini model {model_name}.",
            status=502,
            code="TLS_ERROR",
            retryable=True,
        ) from exc
    except urllib_error.URLError as exc:
        raise ApiError(
            f"Gemini request failed for {model_name}: {exc.reason}",
            status=502,
            code="AI_NETWORK_ERROR",
            retryable=True,
        ) from exc

    text = extract_candidate_text(response_payload)
    return parse_trackers_text(text, current_tracker_count=current_tracker_count)


def parse_trackers_with_gemini(user_input, current_trackers):
    if not user_input.strip():
        raise ApiError("Missing prompt input.", status=400, code="PROMPT_REQUIRED")

    safe_trackers = current_trackers if isinstance(current_trackers, list) else []
    payload = build_payload(user_input, safe_trackers)
    primary_model = get_gemini_primary_model()
    fallback_model = get_gemini_fallback_model()

    try:
        trackers = request_gemini(payload, primary_model, current_tracker_count=len(safe_trackers))
        return {
            "trackers": trackers,
            "model": primary_model,
            "fallback_used": False,
        }
    except ApiError as exc:
        if not exc.retryable or not fallback_model or fallback_model == primary_model:
            raise

    trackers = request_gemini(payload, fallback_model, current_tracker_count=len(safe_trackers))
    return {
        "trackers": trackers,
        "model": fallback_model,
        "fallback_used": True,
    }

File: api/_lib/gemini.js
import { getEnv, requireEnv } from "./runtime.js";

function buildParsePrompt(userInput, currentTrackers) {
  const currentJson = JSON.stringify(currentTrackers, null, 2);
  return `You convert habit-tracker instructions into the final tracker list for a productivity app.

Return only a JSON array. Do not include markdown, code fences, or explanations.

The array must represent the full final tracker list after applying the user's instruction to the current trackers.
If the user wants to add trackers, include them.
If the user wants to rename, modify, or remove trackers, update the array accordingly.
Preserve existing tracker ids when modifying an existing tracker.

Each tracker object should follow these rules:
- id: preserve the existing tracker id when updating an existing tracker, otherwise omit it
- name: short string
- type: one of "binary", "numeric", "session"
- category: one of "study", "diet", "exercise", "habits", "none"
- frequency: "daily" or "weekly"
- logging_mode: "simple" for binary, "quantity" for numeric, "time" for session
- unit: short string, empty string, or null
- goal: number or null
- increments: array of 0 to 3 positive numbers
- primary_action: short verb phrase
- optional_actions: array of up to 3 short strings
- fields: array of field objects with shape { "name": string, "type": "number" | "boolean" | "time", "unit": string | null }

Tracker guidance:
- binary trackers usually use fields like [{"name":"done","type":"boolean","unit":null}]
- numeric trackers usually use one number field
- session trackers usually use [{"name":"duration","type":"time","unit":"minutes"}]
- weekly trackers should still keep the same general shape

Current trackers:
${currentJson}

User instruction:
${userInput}`.trim();
}

function buildResponseSchema() {
  return {
    type: "ARRAY",
    items: {
      type: "OBJECT",
      properties: {
        id: { type: "STRING" },
        name: { type: "STRING" },
        type: { type: "STRING" },
        mode: { type: "STRING" },
        category: { type: "STRING" },
        logging_mode: { type: "STRING" },
        unit: { type: "STRING", nullable: true },
        goal: { type: "NUMBER", nullable: true },
        frequency: { type: "STRING" },
        increments: { type: "ARRAY", items: { type: "NUMBER" } },
        primary_action: { type: "STRING" },
        optional_actions: { type: "ARRAY", items: { type: "STRING" } },
        fields: {
          type: "ARRAY",
          items: {
            type: "OBJECT",
            properties: {
              name: { type: "STRING" },
              type: { type: "STRING" },
              unit: { type: "STRING", nullable: true }
            },
            required: ["name", "type"]
          }
        }
      },
      required: ["name", "type", "category", "logging_mode", "frequency", "increments", "primary_action", "optional_actions", "fields"]
    }
  };
}

function buildGeminiError(message, status = 502) {
  const err = new Error(message);
  err.status = status;
  return err;
}

function extractCandidateText(data) {
  const candidates = Array.isArray(data?.candidates) ? data.candidates : [];
  if (!candidates.length) {
    const blockReason = data?.promptFeedback?.blockReason;
    if (blockReason) {
      throw buildGeminiError(`Gemini blocked the prompt: ${blockReason}`, 422);
    }
    throw buildGeminiError("Gemini returned no candidates");
  }

  const parts = Array.isArray(candidates[0]?.content?.parts) ? candidates[0].content.parts : [];
  const text = parts
    .map((part) => (typeof part?.text === "string" ? part.text : ""))
    .join("")
    .trim();

  if (!text) throw buildGeminiError("Gemini returned an empty response");
  return text;
}

function stripCodeFences(text) {
  return String(text || "")
    .trim()
    .replace(/^```(?:json)?\s*/i, "")
    .replace(/\s*```$/, "")
    .trim();
}

function extractBalancedJsonSnippet(text, openChar, closeChar) {
  const start = text.indexOf(openChar);
  if (start === -1) return "";

  let depth = 0;
  let inString = false;
  let escaped = false;

  for (let index = start; index < text.length; index += 1) {
    const char = text[index];

    if (inString) {
      if (escaped) {
        escaped = false;
      } else if (char === "\\") {
        escaped = true;
      } else if (char === "\"") {
        inString = false;
      }
      continue;
    }

    if (char === "\"") {
      inString = true;
      continue;
    }

    if (char === openChar) depth += 1;
    if (char === closeChar) {
      depth -= 1;
      if (depth === 0) return text.slice(start, index + 1);
    }
  }

  return "";
}

function looksLikeTrackerObject(value) {
  return !!value && typeof value === "object" && !Array.isArray(value) && (
    typeof value.name === "string" ||
    typeof value.type === "string" ||
    Array.isArray(value.fields) ||
    typeof value.frequency === "string"
  );
}

function coerceTrackerArray(value, currentTrackerCount = 0) {
  if (Array.isArray(value)) {
    return value.filter((tracker) => tracker && typeof tracker === "object" && !Array.isArray(tracker));
  }

  if (!value || typeof value !== "object") return null;

  if (Array.isArray(value.trackers)) {
    return value.trackers.filter((tracker) => tracker && typeof tracker === "object" && !Array.isArray(tracker));
  }

  for (const key of ["result", "data", "items", "output"]) {
    if (Array.isArray(value[key])) {
      return value[key].filter((tracker) => tracker && typeof tracker === "object" && !Array.isArray(tracker));
    }
  }

  if (currentTrackerCount === 0 && looksLikeTrackerObject(value)) {
    return [value];
  }

  return null;
}

function summarizeGeminiText(text) {
  return String(text || "")
    .replace(/\s+/g, " ")
    .trim()
    .slice(0, 220);
}

function parseTrackersFromText(text, currentTrackerCount = 0) {
  const cleaned = stripCodeFences(text);
  const candidates = [cleaned];
  const arraySnippet = extractBalancedJsonSnippet(cleaned, "[", "]");
  const objectSnippet = extractBalancedJsonSnippet(cleaned, "{", "}");

  if (arraySnippet && arraySnippet !== cleaned) candidates.push(arraySnippet);
  if (objectSnippet && objectSnippet !== cleaned && objectSnippet !== arraySnippet) candidates.push(objectSnippet);

  for (const candidate of candidates) {
    if (!candidate) continue;
    try {
      let parsed = JSON.parse(candidate);
      if (typeof parsed === "string") {
        parsed = JSON.parse(parsed);
      }
      const trackers = coerceTrackerArray(parsed, currentTrackerCount);
      if (trackers) return trackers;
    } catch {
      // Try the next candidate shape.
    }
  }

  throw buildGeminiError(`Gemini did not return valid JSON array. Received: ${summarizeGeminiText(cleaned)}`);
}

async function requestGemini(payload, modelName, currentTrackerCount = 0) {
  const apiKey = requireEnv("GEMINI_API_KEY", "GOOGLE_API_KEY");
  const url = `https://generativelanguage.googleapis.com/v1beta/models/${modelName}:generateContent`;

  const res = await fetch(url, { method: "POST", headers: { "Content-Type": "application/json", "x-goog-api-key": apiKey }, body: JSON.stringify(payload) });
  const data = await res.json();
  
  if (!res.ok) {
    throw buildGeminiError(data.error?.message || "Gemini API error");
  }

  const text = extractCandidateText(data);
  return parseTrackersFromText(text, currentTrackerCount);
}

export async function parseTrackersWithGemini(userInput, currentTrackers) {
  if (!userInput || !userInput.trim()) {
    const err = new Error("Missing prompt input.");
    err.status = 400;
    throw err;
  }

  const safeTrackers = Array.isArray(currentTrackers) ? currentTrackers : [];
  const payload = { contents: [{ parts: [{ text: buildParsePrompt(userInput.trim(), safeTrackers) }] }], generationConfig: { temperature: 0.2, responseMimeType: "application/json", responseSchema: buildResponseSchema(), maxOutputTokens: 700 } };
  const primaryModel = getEnv("GEMINI_PRIMARY_MODEL", "GEMINI_MODEL") || "gemini-2.5-flash";
  const fallbackModel = getEnv("GEMINI_FALLBACK_MODEL") || "gemini-2.0-flash";
  
  try {
    const trackers = await requestGemini(payload, primaryModel, safeTrackers.length);
    return { trackers, model: primaryModel, fallback_used: false };
  } catch (err) {
    if (!fallbackModel || primaryModel === fallbackModel || err.status !== 502) {
      throw err;
    }
    const trackers = await requestGemini(payload, fallbackModel, safeTrackers.length);
    return { trackers, model: fallbackModel, fallback_used: true };
  }
}
